### PR TITLE
feat: implement cache invalidation api

### PR DESCRIPTION
**Pull request by bkyryliuk**
### SUMMARY
Implements a way to invalidate cached data for the provide datasources.

### TEST PLAN
* unit tests
* local testing

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/cache/api.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+from superset.models.cache import CacheKey
+from superset.views.base_api import BaseSupersetModelRestApi, statsd_metrics
+
+
+class CacheRestApi(BaseSupersetModelRestApi):
+    datamodel = SQLAInterface(CacheKey)
+    resource_name = "cache"
+    allow_browser_login = True
+    class_permission_name = "CacheRestApi"
+
```
Comment (dpgaspar) [inline superset/cache/api.py:37]: use `openapi_spec_component_schemas = CacheInvalidateRequestSchema` to add the schema to the OpenAPI spec

File: superset/cache/api.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            500:
+              $ref: '#/components/responses/500'
+        """
+        try:
+            datasources = CacheInvalidationRequestSchema().load(request.json)
+        except KeyError:
+            return self.response_400(message="Request is incorrect")
+        except ValidationError as error:
+            return self.response_400(
+                message=_("Request is incorrect: %(error)s", error=error.messages)
```
Comment (dpgaspar) [inline superset/cache/api.py:73]: nit: Just use: `return self.response_400(message=error.messages)`

File: tests/superset_test_config.py [hunk 3]
```diff
@@ -76,6 +76,14 @@ def GET_FEATURE_FLAGS_FUNC(ff):
 REDIS_PORT = os.environ.get("REDIS_PORT", "6379")
 REDIS_CELERY_DB = os.environ.get("REDIS_CELERY_DB", 2)
 REDIS_RESULTS_DB = os.environ.get("REDIS_RESULTS_DB", 3)
+REDIS_CACHE_DB = os.environ.get("REDIS_CACHE_DB", 4)
+
+CACHE_CONFIG = {
```
Comment (bkyryliuk) [inline tests/superset_test_config.py:81]: needed for cache to work on CI

**Comment by bkyryliuk** [comment]
@willbarrett , @dpgaspar - ready for review

File: superset/cachekeys/api.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (mistercrunch) [inline superset/cachekeys/api.py:1]: I think the route `/api/v1/cachekey/invalidate` is not ideal. I'm struggling to find something good though

File: superset/cachekeys/api.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    openapi_spec_component_schemas = (CacheInvalidationRequestSchema,)
+
+    @expose("/invalidate", methods=["POST"])
+    @event_logger.log_this
+    @protect()
+    @safe
+    @statsd_metrics
+    def invalidate(self) -> Response:
+        """
+        Takes a list of datasources, finds the associated cache records and
```
Comment (mistercrunch) [inline superset/cachekeys/api.py:46]: Side note (outside the scope of this PR), we'll be renaming datasource to dataset. I think it's mostly done in the UI layer, but has yet to be done in the code.

File: superset/cachekeys/api.py (hunk 4 above)
Comment (bkyryliuk) [reply superset/cachekeys/api.py:1]: agree, I had the same struggles, definitely open to suggestions. 
/api/v1/cache/invalidate could be slightly better

File: superset/cachekeys/api.py (hunk 5 above)
Comment (bkyryliuk) [reply superset/cachekeys/api.py:46]: good catch, I can do it in the separate PR if that works for you, would rename it in API & CacheKey model for consistency.

File: superset/cachekeys/api.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            if ds_obj:
+                datasource_uids.add(ds_obj.uid)
+
+        cache_keys = (
+            db.session.query(CacheKey)
+            .filter(CacheKey.datasource_uid.in_(datasource_uids))
+            .all()
+        )
+        if cache_keys:
+            cache_manager.cache.delete_many(*[c.cache_key for c in cache_keys])
```
Comment (dpgaspar) [inline superset/cachekeys/api.py:94]: A couple of questions:
- if a cache `delete_many` fails, we may end up with delete keys on the cache that still exist on `CacheKey` will a next run of `delete_many` fail if it can't find a key?
- the same applies if a delete `CacheKey` fails, would also be safer to try/except and rollback

File: superset/cachekeys/schemas.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+
+class Datasource(Schema):
+    database_name = fields.String(description="Datasource name",)
+    datasource_name = fields.String(description=datasource_name_description,)
+    schema = fields.String(description="Datasource schema",)
+    datasource_type = fields.String(
+        description=datasource_type_description,
+        validate=validate.OneOf(choices=("druid", "table", "view")),
+        allow_none=True,
```
Comment (dpgaspar) [inline superset/cachekeys/schemas.py:34]: I think that `get_datasource_by_name` will not handle well a `None` here, replace by `required=True`.

File: tests/base_tests.py [hunk 8]
```diff
@@ -69,6 +71,39 @@ def get_resp(
     return resp.data.decode("utf-8")
 
 
+def post_assert_metric(
```
Comment (dpgaspar) [inline tests/base_tests.py:74]: Is this the same but on a different place?

File: tests/cachekeys/api_tests.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    cache_manager.cache.set("cache_key1", "value")
+    cache_manager.cache.set("cache_key2", "value")
+    cache_manager.cache.set("cache_key4", "value")
+    cache_manager.cache.set("cache_keyX", "value")
+
+    rv = post_assert_metric(
+        test_client,
+        "api/v1/cache/invalidate",
+        {
+            "datasource_uids": ["3__druid", "4__druid"],
```
Comment (dpgaspar) [inline tests/cachekeys/api_tests.py:78]: can we add a test with edge case POST's: `{ "datasource_uids": [], "datasources": []}`
`{ "datasource_uids": [], "datasources": [{"datasource_name": "", "datasource_type": null}]}`

etc

**Review by dpgaspar — COMMENTED** [review]
Left a couple of comments

File: tests/base_tests.py (hunk 8 above)
Comment (bkyryliuk) [reply tests/base_tests.py:74]: yes same code, moved it into this function for reuse in the pytests

File: superset/cachekeys/api.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+from superset.cachekeys.schemas import CacheInvalidationRequestSchema
+from superset.connectors.connector_registry import ConnectorRegistry
+from superset.extensions import cache_manager, db, event_logger
+from superset.models.cache import CacheKey
+from superset.views.base_api import BaseSupersetModelRestApi, statsd_metrics
+
+
+class CacheRestApi(BaseSupersetModelRestApi):
+    datamodel = SQLAInterface(CacheKey)
+    resource_name = "cache"
```
[the remark this answers is not in the document]
Comment (bkyryliuk) [reply superset/cachekeys/api.py:38]: renamed to cachekey, I am happy with either way naming it `cachekey` or `cache`
I was thinking about moving it into separate class from the cache key, but most of the instrumentation lives in BaseSupersetModelRestApi that requires to model.

File: tests/cachekeys/api_tests.py (hunk 9 above)
Comment (bkyryliuk) [reply tests/cachekeys/api_tests.py:111]: done, it caught the bug by the way :) thanks !

File: superset/cachekeys/api.py (hunk 6 above)
Comment (bkyryliuk) [reply superset/cachekeys/api.py]: delete_many doesn't fail, only returns true / false if the keys were deleted. e.g. if the key was not found it will return false. Updated CacheKey logic to delete is a single statement & rollback if fails -> that would be more robust.

deleting case & keeping records is not too bad as there still a need to garbage collect those and cache is not a persistent store.

File: superset/cachekeys/schemas.py (hunk 7 above)
Comment (bkyryliuk) [reply superset/cachekeys/schemas.py]: done

**Review by dpgaspar — CHANGES_REQUESTED** [review]
Do you want all the `ModelRestApi` route methods to be available? If not include `include_route_methods = {"invalidate",}` also note this will create a new specific permission `can invalidate on Cache key`.
Check `https://localhost:8080/swagger/v1`

**Comment by bkyryliuk** [comment]
> Do you want all the `ModelRestApi` route methods to be available? If not include `include_route_methods = {"invalidate",}` also note this will create a new specific permission `can invalidate on Cache key`.
> Check `https://localhost:8080/swagger/v1`

addressed, thanks

**Comment by bkyryliuk** [comment]
image

File: superset/cachekeys/api.py (hunk 4 above)
Comment (willbarrett) [reply superset/cachekeys/api.py:1]: How about `DELETE /api/v1/cachekey/<cachekey>`?

**Review by dpgaspar — APPROVED** [review]

File: superset/cachekeys/api.py (hunk 4 above)
Comment (bkyryliuk) [reply superset/cachekeys/api.py:1]: it would be too much struggle to fetch cache keys though the API and then issue deletes for those.
I think we could expose GET / DELETE later once those use cases will arise

**Comment by benceorlai** [comment]
Roadmap: https://github.com/apache-superset/superset-roadmap/issues/74

File: superset/cachekeys/api.py [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+    class_permission_name = "CacheRestApi"
+    include_route_methods = {
+        "invalidate",
+    }
+
+    openapi_spec_component_schemas = (CacheInvalidationRequestSchema,)
+
+    @expose("/invalidate", methods=["POST"])
+    @event_logger.log_this
+    @protect()
```
Comment (ktmud) [inline superset/cachekeys/api.py:49]: It seems this API is not actually protected. Anyone can trigger cache key invalidation with a cURL:

```

```

Is this by design?


It's probably missing a `@permission_name("edit")`.
